### PR TITLE
TUI cursor motion, SGR, and scrolling optimizations

### PR DESCRIPTION
**Pull request by jdebp**
This is #6793, rebased.

### CODE REVIEW

**Comment by justinmk** [comment]
LGTM, @jdebp are you ready for this to be merged?

**Comment by aktau** [comment]
LGTM, although I must admit that I'm not a VT protocol warrior, so my LGTM is worse than the author taking a second look. Would be great if @leonerd could take a look, but if this is tried out in a couple of terminals (tmux, xterm, urxvt, gnome-terminal, iTerm2, Terminal.app). I've tried the latter two + tmux and it seems fine (anything I should pay more attention to?).

**Comment by jdebp** [comment]
iTerm2, Terminal.app are particuarly useful for you to test out @aktau, because they are two that I am unable to test.  Look in particular for scrolling problems with vertically and horizontally split windows, and oddities in changing the cursor shape when entering and leaving insert mode; both with and without screen/tmux in the mix.  With screen/tmux, check that only the current pane is affected by shape changes.

Check over SSH.  Ensure that you transfer the `XTERM_VERSION`, `COLORTERM`, and `KONSOLE_`* environment variables across from client to server.  Also try not doing that, of course.

Or go straight to #6816 and do this.

**Comment by aktau** [comment]
I've tested iTerm2 and Terminal.app locally here as to your suggestions (no tmux, no ssh). The result matrix:

term         | scrolling  | cursor
------       | ---------  | ------
iterm2       | OK         | OK
Terminal.app | **NOT** OK | OK

So, Terminal.app had issues scrolling. I split my window into 4 panes (one in each corner): 

```
| 1 | 2 |
| 3 | 4 |
```

What seemed to happen is that scrolling in pane 1 also scrolled pane 2 (and vice versa). The same for panes 3 and 4. This causes corruption in the screen state. See the screenshots at the end [1].

This is macOS Sierra, and the `$TERM` variable when I'm insude Terminal.app is:

```
13:22:49 neovim/neovim git:(merge-pr-6801) » echo $TERM
xterm-256color
```

[1]: screen shot 2017-05-26 at 13 20 21
screen shot 2017-05-26 at 13 20 47

**Comment by jdebp** [comment]
This is as #6802 leads me to expect.  Now try with #6816.  Your matrix should be success in all cases.

**Comment by aktau** [comment]
> This is as #6802 leads me to expect. Now try with #6816. Your matrix should be success in all cases.

Indeed, Terminal.app works correctly with #6816 (and so does iTerm2). Though I *do* notice sluggishness when scrolling one of the panes in iTerm2 now. Terminal.app seems nice and responsive.

The fact that it's exactly the same Neovim version + configuration makes me think it has something to do with the terminal.

However, I stopwatch timed scrolling to the bottom of the `LICENSE` file and these are the results:

- iTerm2/6816: 10.54s
- iTerm2/master: 10.26s
- iTerm2/6816: 10.27s
- Terminal.app/6816: 10.36s

Basically saying that either I'm imagining things or that Neovim always scrolls equally fast, but doesn't uphold the same framerate while doing so.

Anyway, everything seems good, especially from a functionality point of view.

**Comment by justinmk** [comment]
@aktau That's exactly the same observation I had :) https://github.com/neovim/neovim/pull/6801#issuecomment-304084043

**Comment by aktau** [comment]
@jdebp, any idea about the sluggishness on iTerm2 (but not Terminal.app)? If two people saw it, it may be a real thing. How do we properly debug this? Does your newest rebase change anything there?

Also, do we need to keep this PR open if #6816 seems to subsume it? (Since there's commits in #6816 that fix bugs introduced here, I wouldn't merge it separately in order to have a more bug-free history).

**Comment by jdebp** [comment]
The rebase was just a rebase, and shouldn't have changed anything.  At this point it seems wise to focus on #6816 because that has all of this and a lot more.

I suggest that you try the latest in #6816, and measure there.  We seem to have finally pushed iTerm2 past the highly faulty stage, although I am still unhappy about some of the bizarre output that is in the syscall traces provided by @choco there that I do not yet have a satisfactory explanation for, and about @choco's one (I understand) remaining problem which is backspacing in command editing mode around the edge of the terminal.

I suspect that the two are related, but I have yet to grasp how.  @choco's syscall output looks like malformed multibyte characters of some sort, _following_ a `cnorm` sequence.  I don't see a code path where _anything_ can follow a `cnorm` sequence at the tail of a `write()` system call.

**Comment by justinmk** [comment]
I guess this isn't merge-able, shall we close it?

**Comment by gnachman** [comment]
If you're running into problems with iTerm2, I'd like to know if there are bugs or performance improvements I can make to help you. Especially if there are ways the terminal emulation differs from Terminal that aren't justifiable! Logs from `script` that produce incorrect output ought to be enough for me to track down the issue.

**Comment by justinmk** [comment]
This PR is a subset of #6816, which was merged.
